test-harness/tests/test_cache.py

```diff
@@ -637,3 +638,83 @@ def test_if_failed_request_triggered_update_is_recovered_by_timers(
 
             # Verify that the cache is OK now
             ping_mesos_agent(ar, valid_user_header)
+
+    def test_if_early_boot_stage_is_not_sensitive_to_failures(
+            self, nginx_class, valid_user_header, mocker, log_catcher):
+        # The idea here is to make Backend so slow, that AR is unable to
+        # update cache on first request. AR should then switch to fail-fast
+        # behaviour and wait for timers to update the backend when it recovers.
+
+        refresh_lock_timeout = 10
+        backend_request_timeout = 5
+
+        ar = nginx_class(cache_first_poll_delay=1,
+                         cache_poll_period=3,
+                         cache_expiration=2,
+                         cache_max_age_soft_limit=1200,
+                         cache_max_age_hard_limit=1800,
+                         cache_backend_request_timeout=backend_request_timeout,
+                         cache_refresh_lock_timeout=refresh_lock_timeout,
+                         )
+        agent_id = 'de1baf83-c36c-4d23-9cb0-f89f596cd6ab-S1'
+        url = ar.make_url_from_path('/agent/{}/blah/blah'.format(agent_id))
+        v = Vegeta(log_catcher, target=url, jwt=valid_user_header, rate=3)
+
+        # Make mesos slooowwwww.....
+        # It mus respond slower than backend_request_timeout
+        mocker.send_command(endpoint_id='http://127.0.0.2:5050',
+                            func_name='always_stall',
+                            aux_data=backend_request_timeout * 1.5)
+
+        with GuardedSubprocess(ar):
+            with GuardedSubprocess(v):
+                time.sleep(1)  # let it warm-up!
+
+                # Vegeta by this time should have already clogged AR
```

Comment:
Regarding your first comment:
1. There are a couple of reasons why:
  * I did not want to write new code. Subprocess class is already there, Vegeta is highly configurable. Recreating constant-rate request functionality in Python reliably would take some time.
  * We already had some discussions about GIL, Python, and single-process performance. Using Vegeta seemed future proof - no matter how much more stuff we do in the main Python process, benchmark tool is already a separate process, and we do not have to worry about timing, scheduling, etc... 
  * In this particular case we just wanted to keep the requests hanging on AR, but in future, we may want to benchmark AR components with higher requests-rate. Vegeta is in Go, and we can give AR a nice run for its money this way.
2. It's a metaphor really - imagine that you want to test if your sink is clogged by pouring water at a constant rate and checking if increasing the amount of water that goes from the tap you increase the amount of water that collects in the sink/can't enter the drainage. Here we have something similar - the constant request rate against AR creates a steady "failure state" against we issue our own, extra request and we test how AR handles it. Test this way has to only issue a single request instead of multiple in order to check if we are timing out or 5XX-ing
3. See the last sentence of the last paragraph.

Regarding your second comment:
This is true, I did not say this is the only way ;) The problem I see here is:
a) you need to pay more attention to timing if you want to get reliable test results. If you just flood AR with requests, things get easier I think.
b) you would have to somehow issue the request and then context-switch to `sleep() + issuing another request` while first one is still ongoing. I am sure there is an easy way to do it in Python (something select()-ish?), but I would have to read a bit/google it, nothing comes to my mind immediately.

I hope it helps.